extensions/table.c

```diff
@@ -155,13 +219,49 @@ static table_row *row_from_string(cmark_syntax_extension *self,
       node_cell *cell = (node_cell *)parser->mem->calloc(1, sizeof(*cell));
       cell->buf = cell_buf;
       cell->start_offset = offset;
-      cell->end_offset = offset + cell_matched - 1;
+      if (cell_matched > 0)
+        cell->end_offset = offset + cell_matched - 1;
+      else
+        cell->end_offset = offset;
 
       while (cell->start_offset > 0 && string[cell->start_offset - 1] != '|') {
         --cell->start_offset;
         ++cell->internal_offset;
       }
 
+      cell->cell_data = (node_cell_data *)parser->mem->calloc(1, sizeof(node_cell_data));
+
+      // Check for a column-spanning cell
+      if (row->n_columns > 0 && cmark_strbuf_len(cell->buf) == 0 && cell->start_offset == cell->end_offset) {
+        cell->cell_data->colspan = 0;
+
+        // find the last cell that isn't part of a colspan, and increment that colspan
+        cmark_llist *tmp = row->cells;
+        node_cell *colspan_cell = NULL;
+        while (tmp) {
+          node_cell *this_cell = (node_cell *)tmp->data;
+          if (this_cell->cell_data->colspan > 0)
+            colspan_cell = this_cell;
+          tmp = tmp->next;
+        }
+        if (colspan_cell)
+          ++colspan_cell->cell_data->colspan;
+      } else {
+        cell->cell_data->colspan = 1;
+      }
+
+      // Check for a row-span marker. Actually incrementing the spanning cell's rowspan will happen later
+      cell->cell_data->rowspan = 1;
+      if (parser->options & CMARK_OPT_TABLE_ROWSPAN_DITTO) {
+        if (strcmp(cmark_strbuf_cstr(cell->buf), "\"") == 0) {
+          cell->cell_data->rowspan = 0;
+        }
+      } else {
+        if (strcmp(cmark_strbuf_cstr(cell->buf), "^") == 0) {
```

Comment:
Should declaring row-spans using caret also be opt-in?

Reply:
Right, I'm wondering if it should be so that existing Markdown content isn't affected by this change, e.g., the test in `extensions.txt` below

Reply:
Unfortunately, if i make spans "opt-in" in this way, there's no way to enable that per-test in the spec documents. So my choices are to either not document it in the spec document, or turn them on for *all* the spec tests and have the change to `extensions.txt` regardless. I could see about setting up a new test scaffold just for table spans, but that would complicate the PR more and feel a bit like overkill.

Reply:
@franklinsch I looked into whether i could add a new test file or flag to use to selectively enable table spans on a specific test file, and it will require changing [the python wrapper used in the tests](https://github.com/apple/swift-cmark/blob/gfm/test/cmark.py) to include a hardcoded option constant that is optionally turned on or off. I'm not comfortable adding a copy/pasted constant to the Python code, since i don't know of a way to automatically populate values from C constants in Python. If we still want to have the table spans be fully optional, we effectively cannot document them. (The tests would need to move into `api_test` instead of being hosted in the spec docs.) Since i would like to generate and host the spec document for swift-cmark eventually, it would be unfortunate to completely miss this important feature.

Reply:
In addition, if we add more options like this, i would prefer it if something like https://github.com/apple/swift-markdown/pull/23 landed alongside it so that users of swift-markdown can use more than just what its default settings are.

Reply:
I'm concerned that if we're going to be breaking existing content here. Clients that use this library directly would see a change in behaviour right, at least compared to the upstream fork? This is something that we can enable by default for DocC, though. Or am I missing something?

Reply:
Clients that use Swift-Markdown or Swift-DocC will also see a change in behavior. However, i've already implemented the code change locally to use an option, so i can at least finish the job and move the tests into `api_test`. I'll push up the change today. (I still want [that Swift-Markdown change](https://github.com/apple/swift-markdown/pull/23) to land, though, so people who use Swift-Markdown directly can avoid the content breakage.)